File: lab-1/src/main/java/com/evist0/labforwardlist/model/ForwardList.java
package com.evist0.labforwardlist.model;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;

public class ForwardList<T> implements Serializable, Iterable {
    private class Node implements Serializable {
        private final T value;
        private Node next;

        public Node(T value, Node next) {
            this.value = value;
            this.next = next;
        }

        public Node(T value) {
            this.value = value;
            this.next = null;
        }
    }

    private Node root;

    public void add(T value) {
        if (root == null) {
            root = new Node(value);
            return;
        }

        Node previous = root;

        while (previous.next != null) {
            previous = previous.next;
        }

        previous.next = new Node(value);
    }

    public void insert(int at, T value) {
        if (at == 0) {
            Node next = root;
            root = new Node(value, next);
            return;
        }

        Node previous = root;
        int i = 1;

        while (i != at && previous.next != null) {
            previous = previous.next;
            ++i;
        }

        if (i != at) {
            throw new IndexOutOfBoundsException();
        }

        Node next = previous.next;
        previous.next = new Node(value, next);
    }

    public void remove(int at) {
        if (at == 0) {
            if (root == null) {
                throw new IndexOutOfBoundsException();
            }

            root = root.next;
            return;
        }

        Node previous = root;
        int i = 1;

        while (i != at && previous.next != null) {
            previous = previous.next;
            ++i;
        }

        if (i != at) {
            throw new IndexOutOfBoundsException();
        }

        previous.next = previous.next.next;
    }

    public void sort(Comparator<T> comparator) {
        root = mergeSort(root, comparator);
    }

    public List<T> toList() {
        ArrayList<T> arrayList = new ArrayList<>();

        Node current = root;

        while (current != null) {
            arrayList.add(current.value);
            current = current.next;
        }

        return arrayList;
    }

    public Iterator iterator() {
        return new Iterator() {
            private Node node = root;

            @Override
            public boolean hasNext() {
                return node != null;
            }

            @Override
            public Object next() {
                Object value = node.value;
                node = node.next;
                return value;
            }
        };
    }

    private Node mergeSort(Node h, Comparator<T> comparator) {
        if (h == null || h.next == null) {
            return h;
        }

        Node middle = getMiddle(h);
        Node nextOfMiddle = middle.next;

        middle.next = null;

        Node left = mergeSort(h, comparator);

        Node right = mergeSort(nextOfMiddle, comparator);

        return sortedMerge(left, right, comparator);
    }

    private Node sortedMerge(Node a, Node b, Comparator comparator) {
        Node result;

        if (a == null)
            return b;
        if (b == null)
            return a;

        if (comparator.compare(a.value, b.value) < 0) {
            result = a;
            result.next = sortedMerge(a.next, b, comparator);
        } else {
            result = b;
            result.next = sortedMerge(a, b.next, comparator);
        }
        return result;
    }

    private Node getMiddle(Node head) {
        if (head == null)
            return null;

        Node slow = head, fast = head;

        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }
}

File: lab-2/src/main/scala/com/hamaddo/labforwardlist/model/ForwardList.scala
package com.hamaddo.labforwardlist.model

import java.util.Comparator

class ForwardList extends Iterable[Object] {
  private class Node(val value: Object, var next: Option[Node]) {
    def this(value: Object) = {
      this(value, None)
    }
  }

  private var root: Option[Node] = None

  def add(value: Object): Unit = {
    if (root.isEmpty) {
      root = Option(new Node(value))
      return
    }

    var previous: Node = root.get

    while (previous.next.isDefined) {
      previous = previous.next.get
    }

    previous.next = Option(new Node(value))
  }

  def insert(at: Int, value: Object): Unit = {
    if (at == 0) {
      val next = root
      root = Option(new Node(value, next))
      return
    }

    var previous: Node = root.get
    var i = 1

    while (i != at && previous.next.isDefined) {
      previous = previous.next.get
      i += 1
    }

    if (i != at) {
      throw new IndexOutOfBoundsException
    }

    val next = previous.next
    previous.next = Option(new Node(value, next))
  }

  def remove(at: Int): Unit = {
    if (at == 0) {
      if (root.isEmpty) {
        throw new IndexOutOfBoundsException
      }

      root = root.get.next
      return
    }

    var previous: Node = root.get
    var i = 1

    while (i != at && previous.next.isDefined) {
      previous = previous.next.get
      i += 1
    }

    if (i != at) {
      throw new IndexOutOfBoundsException
    }

    previous.next = previous.next.get.next
  }

  def sort(comparator: Comparator[Object]): Unit = {
    root = mergeSort(root, comparator)
  }

  override def toList: List[Object] = {
    var arrayList: List[Object] = List()

    var current: Option[Node] = root

    while (current.isDefined) {
      arrayList = arrayList :+ current.get.value
      current = current.get.next
    }

    arrayList
  }

  override def iterator: Iterator[Object] = new Iterator[Object]() {
    private var node = root

    override def hasNext: Boolean = node.isDefined

    override def next: Object = {
      val value = node.get.value
      node = node.get.next
      value
    }
  }

  private def mergeSort(h: Option[Node], comparator: Comparator[Object]): Option[Node] = {
    if (h.isEmpty || h.get.next.isEmpty) {
      return h
    }

    val middle = getMiddle(h)
    val nextOfMiddle = middle.get.next

    middle.get.next = None

    val left = mergeSort(h, comparator)

    val right = mergeSort(nextOfMiddle, comparator)

    sortedMerge(left, right, comparator)
  }

  private def sortedMerge(a: Option[Node], b: Option[Node], comparator: Comparator[Object]): Option[Node] = {
    val merged = Option(new Node(None))
    var temp = merged

    var a_ = a
    var b_ = b

    while (a_.isDefined && b_.isDefined) {
      if (comparator.compare(a_.get.value, b_.get.value) < 0) {
        temp.get.next = a_
        a_ = a_.get.next
      }
      else {
        temp.get.next = b_
        b_ = b_.get.next
      }

      temp = temp.get.next
    }

    while (a_.isDefined) {
      temp.get.next = a_
      a_ = a_.get.next
      temp = temp.get.next
    }

    while (b_.isDefined) {
      temp.get.next = b_
      b_ = b_.get.next
      temp = temp.get.next
    }

    merged.get.next
  }

  private def getMiddle(head: Option[Node]): Option[Node] = {
    if (head.isEmpty) {
      return None
    }

    var slow = head
    var fast = head

    while (fast.get.next.isDefined && fast.get.next.get.next.isDefined) {
      slow = slow.get.next
      fast = fast.get.next.get.next
    }

    slow
  }
}